extension/lib/LogHandler.jsm

```diff
@@ -0,0 +1,199 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const { utils: Cu } = Components;
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Timer.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(
+  this, "Config", "resource://pioneer-study-pathfinder/Config.jsm"
+);
+XPCOMUtils.defineLazyModuleGetter(
+  this, "Pioneer", "resource://pioneer-study-pathfinder/lib/Pioneer.jsm"
+);
+XPCOMUtils.defineLazyModuleGetter(
+  this, "PrefUtils", "resource://pioneer-study-pathfinder/lib/PrefUtils.jsm"
+);
+XPCOMUtils.defineLazyServiceGetter(
+  this, "timerManager", "@mozilla.org/updates/timer-manager;1", "nsIUpdateTimerManager"
+);
+
+this.EXPORTED_SYMBOLS = ["LogHandler"];
+
+const UPLOAD_DATE_PREF = "extensions.pioneer-study-pathfinder.lastLogUploadDate";
+
+const TIMER_NAME = "pioneer-pathfinder-timer";
+
+const KILOBYTE = 1024;
+const MEGABYTE = 1024 * KILOBYTE;
+const UPLOAD_LIMIT = 1 * MEGABYTE;
+
+let padding = 0.95;
+let perEntryPingSizeIncrease = {};
+
+let intervalId;
+
+
+this.LogHandler = {
+  startup() {
+    this.uploadPings();
+
+    // Timers done the way they were in the online news study
+    intervalId = setInterval(this.handleInterval.bind(this), Config.logUploadAttemptInterval);
+
+    // Alternate timers to verify if timers were the issue
+    timerManager.registerTimer(
+      TIMER_NAME, this.handleTimer.bind(this), Config.logUploadAttemptInterval
+    );
+  },
+
+  shutdown() {
+    if (intervalId) {
+      clearTimeout(intervalId);
+      timerManager.unregisterTimer(TIMER_NAME);
+    }
+  },
+
+  async handleInterval() {
+    const entries = [{
+      url: "pathfinder",
+      timestamp: Math.round(Date.now() / 1000),
+      details: "intervalFired",
+    }];
+    await Pioneer.utils.submitEncryptedPing("online-news-log", 1, { entries });
+    this.uploadPings("interval");
+  },
+
+  async handleTimer() {
+    const entries = [{
+      url: "pathfinder",
+      timestamp: Math.round(Date.now() / 1000),
+      details: "timerFired",
+    }];
+    await Pioneer.utils.submitEncryptedPing("online-news-log", 1, { entries });
+    this.uploadPings("timer");
+  },
+
+  async generateEntries(type) {
+    const pingCount = Math.floor(Math.random() * 5) + 1; // Returns a random number from 1-5
+
+    const entriesMinSize = UPLOAD_LIMIT * (pingCount - 1);
+
+    const entry = {
+      url: "pathfinder",
+      timestamp: Math.round(Date.now() / 1000),
+      details: `entry:${type}`,
+    };
+
+    // Calculate and cache the size increase of adding one entry to a ping
+    let sizeDelta = perEntryPingSizeIncrease[type];
+    if (!sizeDelta) {
+      const oneEntrySize = Pioneer.utils.getEncryptedPingSize("online-news-log", 1, {
```

Comment:
`getEncryptedPingSize` is async, so you'll need to await this. Otherwise `sizeDelta` becomes `NaN` when you try to subtract two promises.

Reply:
derp! good spot